dev/PushNotifications/PushNotificationsLongRunningTask/platformfactory.cpp

```diff
@@ -0,0 +1,35 @@
+﻿#include "pch.h"
+
+#include <NotificationsLongRunningProcess_h.h>
+
+#include "platform.h"
+#include <wrl/implements.h>
+#include <wrl/module.h>
+#include "platformfactory.h"
+
+using namespace Microsoft::WRL;
+
+HRESULT NotificationsLongRunningProcessFactory::MakeAndInitialize()
+{
+     return S_OK;
+}
+
+IFACEMETHODIMP NotificationsLongRunningProcessFactory::CreateInstance(
+    _In_opt_ IUnknown* /*outer*/,
+    _In_ REFIID /*riid*/,
+    _COM_Outptr_ void** obj)
+{
+    *obj = nullptr;
+
+    std::call_once(m_platformInitializedFlag,
+        [&] {
+            m_platform = Microsoft::WRL::Make<NotificationsLongRunningPlatformImpl>();
+            m_platform->Initialize();            
+        });
+
+    *obj = m_platform.Get();
+    // Add one ref per platform instance request. Deref occurs automatically.
+    m_platform->AddRef();
```

Comment:
Is addref really needed? Refcount will always be 1 since this is a member variable. When a caller gets an instance of platform, isnt the ref-count already 2 without the addref?

Reply:
No. I tested this multiple times. Whoever releases a reference, it reduces the ref count by 1. This is a problem when our platform factory goes down. It tries to free the Platform ComPtr, which is in a bad state at that time:

Platform factory is created
Platform is created -> ref count = 1
COM client CoCreates
We return a ref to the client using Platform.Get() -> ref count is still 1
COM client releases the ref -> ref count  = 0
Shutdown WinMain
PlatformFactory dtor called
Platform ComPtr dtor called. It crashes.

All this issue is avoided using AddRef.